capstone/src/main/java/com/google/step/clients/MetricsClient.java

```diff
@@ -116,125 +99,57 @@ public MetricsClient() {}
+        // Map to maintain a reference to each restaurant and make searching for them easy.
+        // The map and the list share the reference to each restaurantPageViews.
+        Map<String, RestaurantPageViews> restaurantNameMap = new HashMap<>();
+        List<RestaurantPageViews> restaurantPageViewsList = new ArrayList<>();
+        RestaurantPageViews currRestaurant;
 
-        List<Map<String, Object>> clickData = populateClickData(results, firstPageView);
-
-        resultMap.put("firstDate", firstPageView);
-        resultMap.put("clickData", clickData);
-
-        return resultMap;
-    }
-
-    /**
-     * Parses the array of entities and populates a list of maps that contain the required information.
-     * @param entityList list of entities that we need to be turned into a map: List<Entity>
-     * @param firstPageView map of the earliest date in one of the entities: Map<String, Integer>
-     * @return returns a list of maps representative of the click data:
-     *              "clickData":
-     *                  [
-     *                      {
-     *                          "restaurantName": <String>
-     *                          "week": <int>
-     *                          "year": <int>
-     *                          "numClicks": <int>
-     *                      },
-     *                      ...
-     *                  ]
-     */
-    protected List<Map<String, Object>> populateClickData(List<Entity> entityList, Map<String, Integer> firstPageView) {
-        Integer entityYear;
-        Integer entityWeek;
-        Map<String, Object> tempClickData;
-        List<Map<String, Object>> clickData = new ArrayList<>();
-
-        for (Entity entity : entityList) {
-            entityYear = Integer.parseInt(entity.getProperty("year").toString());
-            entityWeek = Integer.parseInt(entity.getProperty("week").toString());
-
-            // Check for an update on the earliest date.
-            firstPageView = checkEarlierDate(firstPageView, entityWeek, entityYear);
-
-            tempClickData = new HashMap<>();
-            tempClickData.put("restaurantName", entity.getProperty("restaurantName").toString());
-            tempClickData.put("week", entityWeek);
-            tempClickData.put("year", entityYear);
-            tempClickData.put("numClicks", entity.getProperty("count"));
-
-            clickData.add(tempClickData);
+        for (Entity entity : results) {
+            String name = entity.getProperty("restaurantName").toString();
+            int year = Integer.parseInt(entity.getProperty("year").toString());
+            int week = Integer.parseInt(entity.getProperty("week").toString());
+            int count = Integer.parseInt(entity.getProperty("count").toString());
+            if (restaurantNameMap.containsKey(name)) {
+                currRestaurant = restaurantNameMap.get(name);
+                currRestaurant.getPageViews().add(new WeeklyPageView(week, year, count));
```

Comment:
This goes with my previous comment above about unmodifiable copy.

It'd be best if you instantiated this list just in here, and then constructed the restaurantPageViews only afterwards.

I know it seems like a waste of memory, but to be fair, you wouldn't be writing a "load all the things" API call in real life. There's ways around the memory issues, but we don't need to go into that now.

The reason I'm harping on about unmodifiable list here is, in general, having unmodifiable (and therefore thread-safe) objects is very valuable, as a general principle. This means you can just document it as thread-safe, and then consumers can go all kinds of wild with concurrency and you can be confident that your stuff will work.

BTW running concurrent operations on non-thread-safe java containers like HashMap and LinkedList can result in hard to reproduce brokenness (e.g. two threads calling "add" on a linked list can result in a bad circular state, and later, someone trying to iterate over the list will get a frozen permanent loop as the next pointer just points back to itself - I've seen this in the real world).

Also note: In contrast to making the list inside the RestaurantPageView unmodifiable by design, it's totally okay to return the `List<RestaurantPageViews>` as a regular modifiable `ArrayList` as you're doing, because you don't retain ownership of that list, so you don't have to defend yourself against what others might do with it. This maximizes flexibility without compromising your own owned API pieces.

Reply:
Yeah that makes a lot of sense. Has been changed!